datastore/postgres/get.go

```diff
@@ -83,72 +82,89 @@ func (s *MatcherStore) Get(ctx context.Context, records []*claircore.IndexRecord
-			rows, err = res.Query()
-			if err != nil {
-				res.Close()
-				return nil, err
+		if vulns != nil { // We already have results we don't need to go back to the DB.
+			if _, ok := vulnSet[rid]; !ok {
+				vulnSet[rid] = make(map[string]struct{})
 			}
-			resCache[rq.query] = rows
-		}
-
-		// unpack all returned rows into claircore.Vulnerability structs
-		for rows.Next() {
-			// fully allocate vuln struct
-			v := &claircore.Vulnerability{
-				Package: &claircore.Package{},
-				Dist:    &claircore.Distribution{},
-				Repo:    &claircore.Repository{},
+			for _, v := range vulns {
+				if _, ok := vulnSet[rid][v.ID]; !ok {
+					vulnSet[rid][v.ID] = struct{}{}
+					results[rid] = append(results[rid], v)
+				}
 			}
-
-			var id int64
-			err := rows.Scan(
-				&id,
-				&v.Name,
-				&v.Description,
-				&v.Issued,
-				&v.Links,
-				&v.Severity,
-				&v.NormalizedSeverity,
-				&v.Package.Name,
-				&v.Package.Version,
-				&v.Package.Module,
-				&v.Package.Arch,
-				&v.Package.Kind,
-				&v.Dist.DID,
-				&v.Dist.Name,
-				&v.Dist.Version,
-				&v.Dist.VersionCodeName,
-				&v.Dist.VersionID,
-				&v.Dist.Arch,
-				&v.Dist.CPE,
-				&v.Dist.PrettyName,
-				&v.ArchOperation,
-				&v.Repo.Name,
-				&v.Repo.Key,
-				&v.Repo.URI,
-				&v.FixedInVersion,
-				&v.Updater,
-			)
-			v.ID = strconv.FormatInt(id, 10)
+			continue
+		}
+		results[rid] = []*claircore.Vulnerability{}
+		err := func() error {
+			rows, err := res.Query()
 			if err != nil {
 				res.Close()
-				return nil, fmt.Errorf("failed to scan vulnerability: %v", err)
+				return fmt.Errorf("error getting rows: %w", err)
 			}
+			defer rows.Close()
+			// unpack all returned rows into claircore.Vulnerability structs
+			for rows.Next() {
+				// fully allocate vuln struct
+				v := &claircore.Vulnerability{
+					Package: &claircore.Package{},
+					Dist:    &claircore.Distribution{},
+					Repo:    &claircore.Repository{},
+				}
 
-			rid := rq.record.Package.ID
-			if _, ok := vulnSet[rid]; !ok {
-				vulnSet[rid] = make(map[string]struct{})
-			}
-			if _, ok := vulnSet[rid][v.ID]; !ok {
-				vulnSet[rid][v.ID] = struct{}{}
-				results[rid] = append(results[rid], v)
+				var id int64
+				err := rows.Scan(
+					&id,
+					&v.Name,
+					&v.Description,
+					&v.Issued,
+					&v.Links,
+					&v.Severity,
+					&v.NormalizedSeverity,
+					&v.Package.Name,
+					&v.Package.Version,
+					&v.Package.Module,
+					&v.Package.Arch,
+					&v.Package.Kind,
+					&v.Dist.DID,
+					&v.Dist.Name,
+					&v.Dist.Version,
+					&v.Dist.VersionCodeName,
+					&v.Dist.VersionID,
+					&v.Dist.Arch,
+					&v.Dist.CPE,
+					&v.Dist.PrettyName,
+					&v.ArchOperation,
+					&v.Repo.Name,
+					&v.Repo.Key,
+					&v.Repo.URI,
+					&v.FixedInVersion,
+					&v.Updater,
+				)
+				v.ID = strconv.FormatInt(id, 10)
+				if err != nil {
+					res.Close()
+					return fmt.Errorf("failed to scan vulnerability: %w", err)
+				}
+
+				if _, ok := vulnSet[rid]; !ok {
```

Comment:
I realize this was here before but I'm wondering: is it possible we see the same `rid` twice in the `rqs` loop and they both have different queries and different sets of vulnerabilities? If so, then it seems like, in the end, we give them both the same set of vulns instead of separate. Is this expected? If so, when?

Reply:
Yes, I think this is a dormant bug, it probably never comes up in practice because:

1. With VEX the repository doesn't matter until `Vulnerable()` is called (so after the querying) so the queries per package are the same.
2. Before that, in the OVAL files, each vulnerability referenced all the possible repos so I suppose you were bound to get a comparable set of Vulns for each package.ID anyway.

I think this probably deserves more discussion and a separate issue.

Reply:
3. Every other time a matcher queries on `driver.RepositoryName`, it's a mono-repository representing that ecosytem

Reply:
So I looked into it a bit, and I realize that's supposed to happen due to the nature of IndexRecords. As in, there is a record for every unique for every repository in every environment in every package. So, there should be an example where we see the same `rid` with different queries (for example two different potential repositories) and it's possible we get two different results. At least in this particular example that I'm thinking of (rpms), this was something which has popped up and was a bug, but I believe switching to VEX resolved it? Or, it's still a problem because we have to guess which repository is responsible for the found rpms (we try them all and see what happens)

Reply:
Yeah, it's not an issue for VEX because the specific repo isn't part of the query anymore, it's checked after the vulns are returned. But, it's still a little bit of a bear trap for future stuff.